src/client/oid.rs

```diff
@@ -0,0 +1,216 @@
+
+    let mut buf: [u8; 4] = [0; 4];
+    BigEndian::write_u32(&mut buf,timestamp);
+    buf
+}
+
+// Generates a new machine id represented as an MD5-hashed 3-byte-encoded hostname string.
+// Represented in Little Endian.
+fn gen_machine_id() -> Result<[u8; 3]> {
+    // Short-circuit if machine id has already been calculated.
+    // Since the generated machine id is not variable, arising race conditions
+    // will have the same machine_bytes result.
+    unsafe {
+        if machine_bytes[0] != 0 || machine_bytes[1] != 0 || machine_bytes[2] != 0 {
+            return Ok(machine_bytes);
+        }
+    }
+
+    // Retrieve hostname through libc
+    let len = 255;
+    let mut buf = Vec::<u8>::with_capacity(len);
+    let ptr = buf.as_mut_ptr();
+    let err = unsafe { gethostname(ptr as *mut libc::c_char, len as libc::size_t) } as i32;
+
+    if err != 0 {
+        return Err(DefaultError("Failed to retrieve hostname for OID generation.".to_owned()))
+    }
+
+    // Count hostname bytes
+    let mut i = -1;
+    let mut byte = 1;
+    while byte != 0 {
+        i += 1;
+        byte = unsafe { *(((ptr as u64) + (i as u64)) as *const u8) };
+    }
+
+    // Convert bytes into string
+    unsafe { buf.set_len(i) };
+    let s = String::from_utf8_lossy(&buf);
+
+    // Hash hostname string
+    let mut md5 = Md5::new();
+    md5.input_str(&s.into_owned()[..]);
+    let hash = md5.result_str();
+
+    // Re-convert string to bytes and grab first three
+    let mut bytes = hash.bytes();
+    let mut vec: [u8; 3] = [0; 3];
+    for i in 0..3 {
+        match bytes.next() {
+            Some(b) => vec[i] = b,
+            None => break,
+        }
+    }
+
+    unsafe { machine_bytes = vec };
+    Ok(vec)
+}
+
+// Gets the process ID and returns it as a 2-byte array.
+// Represented in Little Endian.
+fn gen_pid() -> [u8; 2] {
+    let pid = unsafe { libc::getpid() as u16 };
+    let mut buf: [u8; 2] = [0; 2];
+    LittleEndian::write_u16(&mut buf, pid);
+    buf
+}
+
+// Gets an incremental 3-byte count.
+// Represented in Big Endian.
+fn gen_count() -> Result<[u8; 3]> {
+    // Init oid counter
+    if OID_COUNTER.load(Ordering::SeqCst) == 0 {
+        let mut rng = try!(OsRng::new());
+        let start = rng.gen_range(0, MAX_U24 + 1);
+        OID_COUNTER.store(start, Ordering::SeqCst);
+    }
+
+    let u_counter = OID_COUNTER.fetch_add(1, Ordering::SeqCst);
+
+    // Mod result instead of OID_COUNTER to prevent threading issues.
+    // Static mutexes are currently unstable; once they have been
+    // stabilized, one should be used to access OID_COUNTER and
+    // perform multiple operations atomically.
+    let u = u_counter % MAX_U24;
+
+    // Convert usize to writable u64, then extract the first three bytes.
+    let u_int = u as u64;
+
+    let mut buf: [u8; 8] = [0; 8];
+    BigEndian::write_u64(&mut buf, u_int);
+
+    let buf_u24: [u8; 3] = [buf[0], buf[1], buf[2]];
+    Ok(buf_u24)
+}
+
+#[test]
+fn pid_generation() {    
```

Comment:
These are unit tests that rely on private functions, so they can't be isolated into a separate test module.